File: CodingExercise/ObjectiveC/MasterCollectionViewController.m
//
//  MasterCollectionViewController.m
//  SimpsonsViewerOC
//
//  Created by Samuel Maffei on 7/1/18.
//  Copyright © 2018 sammaffei. All rights reserved.
//

#import "MasterCollectionViewController.h"
#import "DataMgr.h"
#import "Constants.h"
#import "MasterCollectionCells.h"
#import "DetailViewController.h"

@interface MasterCollectionViewController ()

@property (readonly, nonatomic) Boolean haveCompactWidth;

@end

@implementation MasterCollectionViewController

static NSArray *cellIDStrs;

+(void)initialize
    {
    cellIDStrs = @[[TextOnlyColCell cellIndentifier],[IconColCell cellIndentifier]];
    }



-(void)setCurColMode:(CollectionViewMode)newMode
    {
    if (_curColMode != newMode)
        {
        _curColMode = newMode;
            
        if (self.collectionView != nil)
            [self.collectionView reloadData];
        }
    }

-(Boolean)getHaveCompactWidth
    {
    if (self.splitViewController == nil)
        return false;
        
    return self.splitViewController.traitCollection.horizontalSizeClass == UIUserInterfaceSizeClassCompact;
    }

- (void)observeValueForKeyPath:(NSString *)keyPath
                      ofObject:(id)object
                        change:(NSDictionary *)change
                       context:(void *)context
    {
    if ([keyPath isEqualToString:cModelLastUpdatedKeyPath])
        {
        if (self.collectionView != nil)
            [self.collectionView reloadData];
        }
    }

-(void)viewDidLoad
    {
    [super viewDidLoad];
        
    [DataMgr.sharedInstance addDataModelObserver:self];
        
    }

-(void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
    {
    NSString *segueID = [segue identifier];
    UICollectionView *colView = self.collectionView;
        
    if ((segueID == nil) || (colView == nil))
        return;
        
    if ([segueID isEqualToString:@"TextOnlyShowDetail"] ||
        [segueID isEqualToString:@"IconShowDetail"])
        {
        DetailViewController    *controller = (DetailViewController *)((UINavigationController *)segue.destinationViewController).topViewController;
        NSArray<NSIndexPath *>  *selectedItems = [colView indexPathsForSelectedItems];
            
        if ((selectedItems != nil) && (selectedItems.count > 0))
            {
            controller.detailItem = [DataMgr.sharedInstance nthItem:[selectedItems objectAtIndex:0].item];
                
            controller.navigationItem.leftItemsSupplementBackButton = TRUE;
            }
            
        if (self.haveCompactWidth)
            {
            [colView selectItemAtIndexPath:nil animated:TRUE scrollPosition:UICollectionViewScrollPositionNone];
            }
        }
    }

#pragma mark Collection View Methods

-(NSInteger)numberOfSectionsInCollectionView:(UICollectionView *)collectionView
    {
    return 1;
    }

-(NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section
    {
    return [DataMgr.sharedInstance numDataItems];
    }

-(CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout sizeForItemAtIndexPath:(NSIndexPath *)indexPath
    {
    switch (self.curColMode)
        {
        case textOnlyMode:
            return CGSizeMake(collectionView.frame.size.width, 40);
                
        default:
        case iconMode:
            {
            CGFloat side = (collectionView.frame.size.width / 2) - 5;
                
            return CGSizeMake(side, side);
            }
        }
    }

-(UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath
    {
    UICollectionViewCell<SetDataProtocol> *dataCell = [collectionView dequeueReusableCellWithReuseIdentifier:cellIDStrs[self.curColMode]
                                                                                                forIndexPath:indexPath];
    ItemData    *iData = [DataMgr.sharedInstance nthItem:indexPath.item];
        
    if ((iData != nil) && (dataCell != nil))
        {
        [dataCell setData:iData];
            
        return dataCell;
        }
        
    return nil;
    }



@end

File: CodingExercise/Swift/MasterCollectionViewController.swift
//
//  MasterCollectionViewController.swift
//  CodingExercise
//
//  Created by Samuel Maffei on 6/19/18.
//  Copyright © 2018 sammaffei. All rights reserved.
//

import UIKit

class MasterCollectionViewController : UICollectionViewController, UICollectionViewDelegateFlowLayout
    {
    enum CollectionViewMode : Int
        {case textOnly = 0, icon}
    
    
    // what's the current cell type?
    
    var curCellIdentifier : String = TextOnlyColCell.cellIdentifer
    
    var curMode : CollectionViewMode = .textOnly
        {
        didSet
            {
            switch curMode          // set the current cell type based on the mode assignment
                {
                case .textOnly:
                    curCellIdentifier = TextOnlyColCell.cellIdentifer
                case .icon:
                    curCellIdentifier = IconColCell.cellIdentifer
                }
                
                
            self.collectionView?.reloadData()
            }
        }
    
    var detailViewController: DetailViewController? = nil
    
    var haveCompactWidth : Bool
        {
        get
            {
            return self.splitViewController!.traitCollection.horizontalSizeClass == .compact
            }
        }

    
    override func viewDidLoad() {
            super.viewDidLoad()
        
            // add and be notifed if model changes

            DataMgr.sharedInstance.addDataModelObserver(indentifer: String(describing: MasterCollectionViewController.self))
                {
                self.collectionView?.reloadData()
                    
                MBProgressHUD.hide(for: self.view, animated: true) // hide progress hud
                }
        
            if let split = splitViewController {
                let controllers = split.viewControllers
                detailViewController = (controllers[controllers.count-1] as! UINavigationController).topViewController as? DetailViewController
            }
        
            MBProgressHUD.showAdded(to: self.view, animated: true)  // show progress hud
        }
    
    // MARK: - Collection View Stuff
    
    override func numberOfSections(in collectionView: UICollectionView) -> Int
        {
        return 1
        }

    override func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int
        {
        return DataMgr.sharedInstance.dataModelArray.count
        }
    
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize
        {
        // Looked into warning. This appears to be an interface builder bug where it tries to flag you for
        // no fixed sized items (like you may have mad a mistake). I like to remove warnings too, but this one is
        // way off base.
            
        switch curMode
            {
            case .textOnly:
                return CGSize(width: collectionView.frame.width, height: 40.0)
            
            case .icon:
                return CGSize(width: (collectionView.frame.width / 2) - 5, height: (collectionView.frame.width / 2) - 5)
            }
        
        }

    override func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell
        {
        let dataProtocol = collectionView.dequeueReusableCell(withReuseIdentifier: curCellIdentifier, for: indexPath) as! SetDataProtocol
            
        dataProtocol.setData(inData: DataMgr.sharedInstance.dataModelArray[indexPath.item])
        
        return dataProtocol as! UICollectionViewCell
        }
    
    // MARK: - Segues
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        
        switch segue.identifier
            {
            case "TextOnlyShowDetail", "IconShowDetail":
            
                if let indexPaths = collectionView?.indexPathsForSelectedItems
                    {
                    let controller = (segue.destination as! UINavigationController).topViewController as! DetailViewController
                    controller.detailItem = DataMgr.sharedInstance.dataModelArray[indexPaths[0].item]
                    controller.navigationItem.leftBarButtonItem = splitViewController?.displayModeButtonItem
                    controller.navigationItem.leftItemsSupplementBackButton = true
                        
                    // if compact width, we should deselect the item before we segue so it isn't selected when we come back
                        
                    if haveCompactWidth
                        {
                        collectionView?.selectItem(at: nil, animated: true, scrollPosition: [])
                        }

                    }

            
            default:
                break
            }
        }

    }
